src/base/ActionsRouter.sol

```diff
@@ -0,0 +1,143 @@
+    function _executeActions(bytes calldata params) internal isNotLocked {
+        poolManager.unlock(params);
+    }
+
+    /// @notice function that is called by the PoolManager through the SafeCallback.unlockCallback
+    function _unlockCallback(bytes calldata data) internal override returns (bytes memory) {
+        // TODO decode in calldata directly for gas
+        (uint256[] memory actions, bytes[] memory params) = abi.decode(data, (uint256[], bytes[]));
+
+        uint256 numActions = actions.length;
+        if (numActions != params.length) revert LengthMismatch();
+
+        for (uint256 actionIndex = 0; actionIndex < numActions; actionIndex++) {
+            uint256 action = actions[actionIndex];
+
+            // separate actions on a pool, and settling deltas
+            if (action < 0x10) {
+                if (action == Actions.SWAP) _swap(params[actionIndex]);
+                else if (action == Actions.INCREASE_LIQUIDITY) _increaseLiquidity(params[actionIndex]);
+                else if (action == Actions.DECREASE_LIQUIDITY) _decreaseLiquidity(params[actionIndex]);
+                else if (action == Actions.MINT_POSITION) _mintPosition(params[actionIndex]);
+                else if (action == Actions.BURN_POSITION) _burnPosition(params[actionIndex]);
+                else if (action == Actions.DONATE) _donate(params[actionIndex]);
+                else _handleAdditionalActions1(action, params[actionIndex]);
+            } else {
+                if (action == Actions.SETTLE) _settle(params[actionIndex]);
+                else if (action == Actions.TAKE) _take(params[actionIndex]);
+                else if (action == Actions.MINT_6909) _mint6909(params[actionIndex]);
+                else if (action == Actions.BURN_6909) _burn6909(params[actionIndex]);
+                else if (action == Actions.CLEAR_DELTA) _clearDelta(params[actionIndex]);
+                else if (action == Actions.SWEEP_ETH) _sweepETH(params[actionIndex]);
+                else _handleAdditionalActions2(action, params[actionIndex]);
+            }
+        }
+
+        // TODO do we want to return anything?
+    }
+
+    /// @notice function that returns the msg.sender who called executeActions
+    function _msgSender() internal view virtual returns (address) {
+        return Locker.get();
+    }
+
+    /// @notice function to settle the open delta of a currency
+    /// @dev The `_pay` function must implement necessary safety checks on the value of payer.
+    /// If approvals are used on this contract, do not allow `_pay` to pull tokens from any payer.
+    /// @dev The function is virtual and can be overrided if different parameters or logic are required.
+    function _settle(bytes memory params) internal virtual {
+        // TODO decode in calldata
+        (Currency currency, address payer) = abi.decode(params, (Currency, address));
+
+        int256 delta = poolManager.currencyDelta(address(this), currency);
```

Comment:
I feel like the base contract should have no impl? Or what was the reasoning to have some impls for some functions and not others?

Reply:
My reasoning was just that theres only really 1 way youre ever going to take/settle theres no variation there. The variation in `settle` comes down to the `pay` function which is abstract still

Reply:
It also enforces the `sync-pay-settle` pattern nicely

Reply:
But can delete if youd rather